### ISSUE TITLE
When using Apple MPS, the bias of BatchNorm1d becomes extremely large.

### ISSUE THREAD

**Issue by mnm1001** [pytorch/pytorch#156555]
### 🐛 Describe the bug

When using Apple MPS, building a model structured as BatchNorm1d -> LSTM -> BatchNorm1d leads to the bias of the first BatchNorm1d layer becoming extremely large during training. However, the values remain normal when using the CPU.

code example:
```
import torch
import torch.nn as nn
import pandas as pd
torch.manual_seed(1)
import numpy as np

def analyze_gradients(model, shouldBackward=False):
    layer_stats = {}
    norms_per_layer = []

    for name, param in model.named_parameters():
        if param.grad is None:
            continue

        grad = param.grad.data
        grad_norm = torch.norm(grad).item()
        grad_min = grad.min().item()
        grad_max = grad.max().item()
        grad_mean = grad.mean().item()

        param_data = param.data
        param_norm = torch.norm(param_data).item()
        param_min = param_data.min().item()
        param_max = param_data.max().item()
        param_mean = param_data.mean().item()

        layer_stats[name] = {
            "grad": {
                "norm": grad_norm,
                "min": grad_min,
                "max": grad_max,
                "mean": grad_mean,
            },
            "param": {
                "norm": param_norm,
                "min": param_min,
                "max": param_max,
                "mean": param_mean,
            },
        }
        norms_per_layer.append(grad_norm)

    sorted_stats = sorted(
        layer_stats.items(), key=lambda x: x[1]["grad"]["norm"], reverse=True
    )

    grad_data = []
    for name, stats in sorted_stats:
        grad_data.append(
            {
                "Layer": name,
                "Grad Norm": f"{stats['grad']['norm']:.2f}",
                "Grad Mean": f"{stats['grad']['mean']:.2f}",
                "Grad Min": f"{stats['grad']['min']:.2f}",
                "Grad Max": f"{stats['grad']['max']:.2f}",
                "Param Norm": f"{stats['param']['norm']:.2f}",
                "Param mean": f"{stats['param']['mean']:.2f}",
                "Param min": f"{stats['param']['min']:.2f}",
                "Param max": f"{stats['param']['max']:.2f}",
            }
        )

    df = pd.DataFrame(grad_data)
    print(df.to_string(index=False))

    avg_norm = np.mean(norms_per_layer)
    print(f"\avg_norm: {avg_norm:.2f}")

    return


class BN_LSTM_BN(nn.Module):
    def __init__(self, input_size, hidden_size, num_layers=1):
        super(BN_LSTM_BN, self).__init__()
        self.bn1 = nn.BatchNorm1d(input_size)
        self.lstm = nn.LSTM(
            input_size=input_size,
            hidden_size=hidden_size,
            num_layers=num_layers,
            batch_first=True,
        )
        self.bn2 = nn.BatchNorm1d(hidden_size)

    def forward(self, x):
        bn1_out = x.permute(0, 2, 1)
        bn1_out = self.bn1(bn1_out)
        bn1_out = bn1_out.permute(0, 2, 1)
        lstm_out, _ = self.lstm(bn1_out)
        last_step_out = lstm_out[:, -1, :]
        output = self.bn2(last_step_out)
        return output


device = "mps"
device = "cpu"


def test_model():
    batch_size = 2
    seq_len = 2
    input_size = 3
    hidden_size = 33
    x = torch.randn(batch_size, seq_len, input_size)
    x = torch.Tensor(
        [
            [
                [0.5594822, 0.11856875, 0.47782624],
                [0.4196736, 0.15617272, 0.42347646],
            ],
            [
                [0.5625423, 0.12889613, 0.58401215],
                [0.4435455, 0.19877888, 0.4796756],
            ],
        ]
    ).to(
        device
    )
    model = BN_LSTM_BN(input_size, hidden_size).to(device)
    output = model(x)
    output.mean().backward()
    analyze_gradients(model)
    y_pred = model(x)
    y_pred.mean().backward()
    analyze_gradients(model)
    print("Input shape:", x.shape)
    print("Output shape:", output.shape)
    print("\nOutput sample:")
    print(output[0, :5])
    return model, output


if __name__ == "__main__":
    model, output = test_model()
```

mps result:

![Image](https://github.com/user-attachments/assets/04ada31b-800d-492c-9382-bfd252eafaeb)

cpu result:

![Image](https://github.com/user-attachments/assets/fa97200d-dc34-448e-87de-151ea2d7e3e8)

### Versions

PyTorch version: 2.7.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 15.4.1 (arm64)
GCC version: Could not collect
Clang version: 17.0.0 (clang-1700.0.13.3)
CMake version: Could not collect
Libc version: N/A

Python version: 3.13.5 (main, Jun 11 2025, 15:36:57) [Clang 17.0.0 (clang-1700.0.13.3)] (64-bit runtime)
Python platform: macOS-15.4.1-arm64-arm-64bit-Mach-O
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] mypy_extensions==1.1.0
[pip3] numpy==1.26.4
[pip3] torch==2.7.0
[pip3] torchvision==0.22.0
[conda] Could not collect

**Comment by Isalia20** [comment]
I'll take a look at this
